Source/Core/VideoCommon/Resources/CustomResourceManager.cpp

```diff
@@ -0,0 +1,231 @@
+
+  m_xfb_event =
+      AfterFrameEvent::Register([this](Core::System&) { XFBTriggered(); }, "CustomResourceManager");
+
+  m_invalid_array_texture = CreateInvalidArrayTexture();
+  m_invalid_color_texture = CreateInvalidColorTexture();
+  m_invalid_cubemap_texture = CreateInvalidCubemapTexture();
+  m_invalid_transparent_texture = CreateInvalidTransparentTexture();
+}
+
+void CustomResourceManager::Shutdown()
+{
+  if (m_async_shader_compiler)
+    m_async_shader_compiler->StopWorkerThreads();
+
+  m_asset_cache.Shutdown();
+  m_worker_thread.Shutdown();
+  Reset();
+}
+
+void CustomResourceManager::Reset()
+{
+  m_material_resources.clear();
+  m_shader_resources.clear();
+  m_texture_data_resources.clear();
+  m_texture_sampler_resources.clear();
+
+  m_invalid_transparent_texture.reset();
+  m_invalid_color_texture.reset();
+  m_invalid_cubemap_texture.reset();
+  m_invalid_array_texture.reset();
+
+  m_asset_cache.Reset();
+  m_texture_pool.Reset();
+  m_worker_thread.Reset("resource-worker");
+}
+
+void CustomResourceManager::MarkAssetDirty(const CustomAssetLibrary::AssetID& asset_id)
+{
+  m_asset_cache.MarkAssetDirty(asset_id);
+}
+
+void CustomResourceManager::XFBTriggered()
+{
+  m_asset_cache.Update();
+}
+
+void CustomResourceManager::SetHostConfig(const ShaderHostConfig& host_config)
+{
+  for (auto& [id, shader_resources] : m_shader_resources)
+  {
+    for (auto& [key, shader_resource] : shader_resources)
+    {
+      shader_resource->SetHostConfig(host_config);
+
+      // Hack to get access to resource internals
+      Resource* resource = shader_resource.get();
+
+      // Tell shader and references to trigger a reload
+      // on next usage
+      resource->NotifyAssetChanged(false);
+    }
+  }
+
+  m_host_config.bits = host_config.bits;
+}
+
+TextureDataResource* CustomResourceManager::GetTextureDataFromAsset(
+    const CustomAssetLibrary::AssetID& asset_id,
+    std::shared_ptr<VideoCommon::CustomAssetLibrary> library)
+{
+  const auto [it, added] = m_texture_data_resources.try_emplace(asset_id, nullptr);
+  if (added)
+  {
+    it->second = std::make_unique<TextureDataResource>(CreateResourceContext(asset_id, library));
+  }
+  ProcessResource(it->second.get());
+  return it->second.get();
+}
+
+MaterialResource* CustomResourceManager::GetMaterialFromAsset(
+    const CustomAssetLibrary::AssetID& asset_id, const GXPipelineUid& pipeline_uid,
+    std::shared_ptr<VideoCommon::CustomAssetLibrary> library)
+{
+  const auto [it_container, added_container] =
+      m_material_resources.try_emplace(asset_id, PipelineIdToMaterial{});
+  const auto [it_uid, added_uid] =
+      it_container->second.try_emplace(PipelineToHash(pipeline_uid), nullptr);
+  if (added_uid)
+  {
+    it_uid->second =
+        std::make_unique<MaterialResource>(CreateResourceContext(asset_id, library), pipeline_uid);
+  }
+  ProcessResource(it_uid->second.get());
+  return it_uid->second.get();
```

Comment:
```suggestion
  auto& resource = m_material_resources[asset_id][PipelineToHash(pipeline_uid)];
  if (resource == nullptr)
  {
    resource =
        std::make_unique<MaterialResource>(CreateResourceContext(asset_id, library), pipeline_uid);
  }
  ProcessResource(resource.get());
  return resource.get();
```
There's a few other occurrences of this pattern, where I think `try_emplace` is overly verbose compared to `operator[]`